Fixed setting some state flags in wxNotebook::HitTest (wxMSW)

a-wi opened the pull request:
State represented by TCHT_ONITEM Win API flag is not a superposition TCHT_ONITEMICON and TCHT_ONITLABEL states but it represents a separate state. The same applies to wxBK_HITTEST_xxx flags where state represented by wxBK_HITTEST_ONITEM is not a superposition of wxBK_HITTEST_ONICON and wxBK_HITTEST_ONLABEL.

In the current implementation it can be observed (in the notebook sample) that pointing over a tab item but not over its icon or label is reported as wxBK_HITTEST_ONICON | wxBK_HITTEST_ONLABEL.

vadz commented on include/wx/bookctrl.h:
```diff
@@ -38,7 +38,7 @@ enum
     wxBK_HITTEST_ONICON  = 2,   // on icon
     wxBK_HITTEST_ONLABEL = 4,   // on label
-    wxBK_HITTEST_ONITEM  = wxBK_HITTEST_ONICON | wxBK_HITTEST_ONLABEL,
+    wxBK_HITTEST_ONITEM  = 16,  // on tab control but not on its icon or label
```
These values don't actually have to be bitmasks, do they? I think we should either leave this one as 6 or renumber all of them to go from 0 to 4

vadz commented:
I am actually not sure at all how is this supposed to work to be honest... The goal probably should be to allow the code to check whether the point is over the icon or the label if supported by the platform (MSW and GTK) but still also check if it's over any part of the tab at all under the others (OS X). I don't understand why do we return `NOWHERE | ONPAGE` for the points over the client area, it looks like just `ONPAGE` should be enough actually.

Do you think we should continue using the bitmasks or not?

a-wi commented:
Indeed, it seems that implementation and interpretation of wxBK_HITTEST_* flags is a bit inconsistent. According to the documentation wxBK_HITTEST_[ONITEM | ONICON | ONLABEL] (1) are mutually exclusive states but wxBK_HITTEST_[NOWHERE | ONPAGE] (2) are not.
In the current wxGTK's HitTest implementation, flags (1) are treated as mutually exclusive ones, flags (2) can coexist.
I think that it would be best to consider all these states as mutually exclusive ones (and number them naturally from 0 to n) but perhaps for backward compatibility reasons they would need to remain as bitmasks.

vadz commented:
From compatibility point of view, I guess it would indeed be better to continue allowing treating them as bitmasks, so that the code written to do `if ( ht & (wxBK_HITTEST_ONITEM | wxBK_HITTEST_ONICON))` would continue to work. But on the implementation side we probably should make it clear that only one of these bits will ever be set.

What do you think?

a-wi committed:
Fixed setting wxBK_HITTEST_xxx flags in wxNotebook::HitTest.

The state represented by TCHT_ONITEM Win API flag is not a superposition TCHT_ONITEMICON and TCHT_ONITLABEL states but it represents a separate state. The fact that binary value of TCHT_ONITEM is bitwise-OR operation on TCHT_ONITEMICON and TCHT_ONITEMLABEL doesn't matter here.
The same applies to wxBK_HITTEST_xxx flags where state represented by wxBK_HITTEST_ONITEM is not a superposition of wxBK_HITTEST_ONICON and wxBK_HITTEST_ONLABEL.

Add note to wxBookCtrl::HitTest documentation that wxBK_HITTEST_ONICON, wxBK_HITTEST_ONLABEL, wxBK_HITTEST_ONITEM are mutually exclusive bits.

a-wi committed:
Check wxBK_HITTEST_ONITEM state in hit test in notebook sample.

a-wi commented:
I agree that for backward compatibility reasons these flags should remain as bitmasks.
I updated wxBookCtrl::HitTest documentation with note that some bits are mutually exclusive.
